### PR TITLE
Add test cases for newly added functions in Internal.Data.Fold

### LINKED ISSUE #675: Need tests for folds
The following APIs which are exposed to the users do not have any tests:

* drain
* drainBy
* mean
* variance
* stdDev
* mconcat
* foldMap
* foldMapM
* lookup
* sequence
* mapM
* tee
* distribute
* partition
* unzip

1) We need tests for these even if it is a trivial test. Without tests, we won't even know if an operation got accidentally removed during refactor. The recent Fold changes are all pervasive and I am concerned that it might break some of these routines which do not have any tests.

2) The test file for folds is not properly formatted, we need to format it.

3) Several routines in `Streamly.Internal.Data.Fold` also require tests.

### PR DESCRIPTION
**Pull request by rnjtranjan**
Fix #675

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fix #739

File: test/Streamly/Test/Data/Fold.hs [hunk 1]
```diff
@@ -179,7 +184,8 @@ elemIndex elm ls = do
             let fld = S.fold (FL.any (== elm)) (S.fromList ls)
             in fld `shouldReturn` False
         Just idx ->
-            let fld = S.fold (FL.any (== elm)) (S.fromList $ Prelude.take idx ls)
+            let fld = S.fold (FL.any (== elm))
+                        (S.fromList $ Prelude.take idx ls)
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:188]:
```
            let fld =
                    S.fold (FL.any (== elm)) (S.fromList $ Prelude.take idx ls)
```

File: test/Streamly/Test/Data/Fold.hs [hunk 2]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+    where
+
+    action ls = do
+        let f = \x -> if odd x then Left x else Right x
+        v1 <- run
+                $ S.fold (F.partitionBy f FL.length FL.length)
+                $ S.fromList ls
+        let v2 = foldl (\b a -> if odd a then b+1 else b) 0 ls
+            v3 = foldl (\b a -> if even a then b+1 else b) 0 ls
+        assert (v1 == (v2, v3))
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs]: This is implemented in terms of `partitionByM` so this test can be skipped.

File: test/Streamly/Test/Data/Fold.hs [hunk 3]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+    forAll (listOf1 (chooseInt (intMin, intMax)))
+        $ \ls0 -> monadicIO $ action ls0
+
+    where
+
+    action ls = do
+        let f = \x -> if odd x then return (Left x) else return (Right x)
+        v1 <- run $ S.fold (F.partitionByM f FL.length FL.length) $ S.fromList ls
+        let v2 = foldl (\b a -> if odd a then b+1 else b) 0 ls
+            v3 = foldl (\b a -> if even a then b+1 else b) 0 ls
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:466]: You could `length $ filter odd`

File: test/Streamly/Test/Data/Fold.hs [hunk 4]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+demux :: Expectation
+demux =
+    let table = Data.Map.fromList [("SUM", F.sum), ("PRODUCT", F.product)]
+        input = Stream.fromList
+                [ ("SUM", 1::Int)
+                , ("PRODUCT", 2::Int)
+                , ("SUM",3)
+                , ("PRODUCT", 4::Int)
+                ]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:493]: Add a type signature to the entire list instead? `::Int` on one element is enough.

Comment (adithyaov) [reply test/Streamly/Test/Data/Fold.hs:493]: Indentation

File: test/Streamly/Test/Data/Fold.hs [hunk 5]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+    in Stream.fold (F.demux table) input
+        `shouldReturn`
+        Data.Map.fromList [("PRODUCT", 8),("SUM", 4)]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:496]: Indentation

File: test/Streamly/Test/Data/Fold.hs [hunk 6]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+demuxWithSum :: Expectation
+demuxWithSum =
+    let f x = ("SUM", x::Int)
+        table = Data.Map.fromList [("SUM", F.sum)]
+        input = Stream.fromList [1, 4]
+
+    in Stream.fold (F.demuxWith f table) input
+        `shouldReturn`
+        Data.Map.fromList [("SUM", 5)]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:507]: zzzzIndentation

File: test/Streamly/Test/Data/Fold.hs [hunk 7]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+demuxWithProduct :: Expectation
+demuxWithProduct =
+    let f x = ("PRODUCT", x::Int)
+        table = Data.Map.fromList [("PRODUCT", F.product)]
+        input = Stream.fromList [2, 4]
+
+    in Stream.fold (F.demuxWith f table) input
+        `shouldReturn`
+        Data.Map.fromList [("PRODUCT", 8)]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:517]: Why so many tests for `demuxWith`?

File: test/Streamly/Test/Data/Fold.hs [hunk 8]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+demuxDefaultWithSum :: Expectation
+demuxDefaultWithSum =
+    let f x = ("PRODUCT", x::Int)
+        table = Data.Map.fromList [("PRODUCT", FL.product)]
+        input = Stream.fromList [2, 4]
+
+    in Stream.fold (F.demuxDefaultWith f table (F.map snd FL.sum)) input
+        `shouldReturn`
+        (Data.Map.fromList [("PRODUCT" , 8)] , 0)
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:527]: The function name has `Sum` but the definition tests for `PRODUCT`.

File: test/Streamly/Test/Data/Fold.hs [hunk 9]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+demuxDefaultWithProduct :: Expectation
+demuxDefaultWithProduct =
+    let f x = ("PRODUCT", x::Int)
+        table = Data.Map.fromList [("PRODUCT", FL.product)]
+        input = Stream.fromList [2, 4]
+
+    in Stream.fold (F.demuxDefaultWith f table (F.map snd FL.product)) input
+        `shouldReturn`
+        (Data.Map.fromList [("PRODUCT" , 8)] , 1)
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:537]: For this, you should test for 2 cases,
1. When the table contains the default cases
2. When it does not
3. Basically test the default and the table

Comment (adithyaov) [reply test/Streamly/Test/Data/Fold.hs:537]: Also, indentation.

File: test/Streamly/Test/Data/Fold.hs (hunk 8 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Fold.hs:527]: Indentation

File: test/Streamly/Test/Data/Fold.hs [hunk 10]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+demuxDefault :: Expectation
+demuxDefault =
+    let table =  Data.Map.fromList [("SUM", F.sum), ("PRODUCT", F.product)]
+        input = Stream.fromList
+            [ ("SUM", 1::Int)
+            , ("PRODUCT", 2::Int)
+            , ("SUM",3)
+            , ("PRODUCT", 4::Int)
+            ]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:547]: 1. Indentation
2. Some comments same as previously stated

File: test/Streamly/Test/Data/Fold.hs [hunk 11]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+    in Stream.fold (F.demuxDefault table (F.map snd FL.product)) input
+        `shouldReturn`
+        (Data.Map.fromList [("PRODUCT", 8), ("SUM", 4)], 1)
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:550]: Indentation

File: test/Streamly/Test/Data/Fold.hs (hunk 6 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Fold.hs:507]: Test for case when table does not have the key.

File: test/Streamly/Test/Data/Fold.hs [hunk 12]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+classifyWith :: Expectation
+classifyWith =
+    let input = Stream.fromList [("ONE",1),("ONE",1.1),("TWO",2), ("TWO",2.2)]
+    in Stream.fold (F.classifyWith fst (F.map snd F.toList)) input
+        `shouldReturn`
+        Data.Map.fromList [("ONE",[1.0, 1.1 :: Double]), ("TWO",[2.0, 2.2])]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:557]: Indentation

File: test/Streamly/Test/Data/Fold.hs [hunk 13]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+            Stream.fromList
+            [
+              ("ONE", (1::Int, 1))
+            , ("ONE", (1, 1.1:: Double))
+            , ("TWO", (2, 2))
+            , ("TWO",(2, 2.2))
+            ]
+    in Stream.fold (F.classify (F.map snd F.toList)) input
+        `shouldReturn`
+        Data.Map.fromList [("ONE",[1.0, 1.1 :: Double]), ("TWO",[2.0, 2.2])]
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:571]: Indentation

File: test/Streamly/Test/Data/Fold.hs [hunk 14]
```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
...
+
+splitAt :: Expectation
+splitAt =
+    Stream.fold
+    (F.splitAt 6 F.toList F.toList)
+    (Stream.fromList "Hello World!")
+    `shouldReturn` ("Hello ","World!")
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:578]: Indentation

File: test/Streamly/Test/Data/Fold.hs [hunk 15]
```diff
@@ -518,9 +660,12 @@ main = hspec $ do
 
         -- Partitioning
         prop "partition" Main.partition
+        prop "partitionBy" partitionBy
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs:665]: Not required. We are testing for `partitionByM`

Comment (adithyaov) [reply test/Streamly/Test/Data/Fold.hs:665]: Please check if `partitionBy` is represented in terms of `partitionByM` before removing this.

File: test/Streamly/Test/Data/Fold.hs [hunk 16]
```diff
@@ -459,7 +459,9 @@ partitionByM =
 
     action ls = do
         let f = \x -> if odd x then return (Left x) else return (Right x)
-        v1 <- run $ S.fold (F.partitionByM f FL.length FL.length) $ S.fromList ls
+        v1 <- run 
+            $ S.fold (F.partitionByM f FL.length FL.length) 
+            $ S.fromList ls
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs]: Incorrect formatting.
Either
```
        v1 <-
            run $ S.fold (F.partitionByM f FL.length FL.length) $ S.fromList ls
```
or,
```
        v1 <-
            run
                $ S.fold (F.partitionByM f FL.length FL.length)
                $ S.fromList ls
```

File: test/Streamly/Test/Data/Fold.hs [hunk 17]
```diff
@@ -474,8 +476,8 @@ partitionBy =
     action ls = do
         let f = \x -> if odd x then Left x else Right x
         v1 <- run
-                $ S.fold (F.partitionBy f FL.length FL.length)
-                $ S.fromList ls
+            $ S.fold (F.partitionBy f FL.length FL.length)
+            $ S.fromList ls
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Fold.hs]: Indent is required.

File: test/Streamly/Test/Data/Fold.hs (hunk 7 above)
Comment (rnjtranjan) [reply test/Streamly/Test/Data/Fold.hs:517]: due to all different combinations.

File: test/Streamly/Test/Data/Fold.hs (hunk 9 above)
Comment (rnjtranjan) [reply test/Streamly/Test/Data/Fold.hs:537]: There is demuxWithProduct for this.

**Review by harendra-kumar — COMMENTED** [review]
Indentation/formatting can be improved in some function calls.

**Review by harendra-kumar — COMMENTED** [review]
1. `Data.Fold.map` should be replaced by `Data.Fold.lmap`.  
2. Try to use non-internal imports as much as possible.

**Comment by harendra-kumar** [comment]
Also, please squash the commits.

**Pull request merged**